server/state_test.go

```diff
@@ -545,3 +555,555 @@ func BenchmarkCallStateClone(b *testing.B) {
 		})
 	}
 }
+
+func TestCleanUpState(t *testing.T) {
+	mockAPI := &pluginMocks.MockAPI{}
+	mockMetrics := &serverMocks.MockMetrics{}
+
+	p := Plugin{
+		MattermostPlugin: plugin.MattermostPlugin{
+			API: mockAPI,
+		},
+		metrics:           mockMetrics,
+		callsClusterLocks: map[string]*cluster.Mutex{},
+	}
+
+	store, tearDown := NewTestStore(t)
+	t.Cleanup(tearDown)
+	p.store = store
+
+	t.Run("plugin mode", func(t *testing.T) {
+		defer mockAPI.AssertExpectations(t)
+		defer mockMetrics.AssertExpectations(t)
+
+		t.Run("no calls", func(t *testing.T) {
+			mockAPI.On("LogDebug", "cleaning up calls state",
+				"origin", mock.AnythingOfType("string")).Once()
+
+			err := p.cleanUpState()
+			require.NoError(t, err)
+		})
+
+		t.Run("ongoing calls", func(t *testing.T) {
+			defer ResetTestStore(t, p.store)
+
+			channelID := model.NewId()
+			postID := model.NewId()
+			userID := model.NewId()
+			callID := model.NewId()
+
+			call := &public.Call{
+				ID:        callID,
+				CreateAt:  time.Now().UnixMilli(),
+				ChannelID: channelID,
+				StartAt:   time.Now().UnixMilli(),
+				PostID:    postID,
+				ThreadID:  model.NewId(),
+				OwnerID:   userID,
+			}
+			err := p.store.CreateCall(call)
+			require.NoError(t, err)
+
+			createPost(t, store, postID, userID, channelID)
+
+			err = p.store.CreateCallSession(&public.CallSession{
+				ID:     "connA",
+				CallID: callID,
+				UserID: "userA",
+				JoinAt: time.Now().UnixMilli(),
+			})
+			require.NoError(t, err)
+
+			mockAPI.On("LogDebug", "cleaning up calls state",
```

Comment:
this must have been a pain to get exactly right, but I kind of like that the tests have the logs laid out like this -- you can see exactly what's supposed to happen. It's easy to understand.

Reply:
Yeah, it takes some time, but it ensures the tests are solid, even in the presence of heavy mocking.